lld/test/wasm/custom-section-name.ll

```diff
@@ -1,4 +1,4 @@
-; RUN: llc -filetype=obj %s -o %t.o
+; RUN: llc -filetype=obj -mattr=-bulk-memory %s -o %t.o
```

Comment:
Why do we need to disable bulk memory here and in many other files?

Reply:
In lld/test/wasm/custom-section-name.ll, it's because bulk-memory makes the target_features section bigger and thus changes some of the section offsets and it was easier to just disable bulk-memory than to figure out the new section offsets. I can update the section offsets if you prefer though.

In lld/test/wasm/data-segments.ll, it's because the test is written to test both bulk-memory and non-bulk-memory, and to keep it doing that, it's now necessary to explicitly disable bulk-memory for the non-bulk-memory case.

In lld/test/wasm/lto/stub-library-libcall.s it's because the test is testing the use of a memcpy stub function, and with bulk-memory it gets inlined.

llvm/test/MC/WebAssembly/extern-functype-intrinsic.ll is similarly testing memset calls.

llvm/test/MC/WebAssembly/libcall.ll is similarly testing memcpy calls.

Reply:
Thanks for the explanation!

Reply:
I've now investigated custom-section-name.ll more.

With bulk-memory, we get a `__wasm_init_memory` function:
```
(module $custom-section-name.bss.wasm
  (type (;0;) (func))
  (import "env" "memory" (memory (;0;) 2))
  (func $__wasm_call_ctors (;0;) (type 0))
  (func $__wasm_init_memory (;1;) (type 0)
    i32.const 1036
    i32.const 0
    i32.const 4
    memory.fill
  )
  (global $__stack_pointer (;0;) (mut i32) i32.const 66576)
  (start $__wasm_init_memory)
  (data $WowZero! (;0;) (i32.const 1024) "\00\00\00\00")
  (data $MyAwesomeSection (;1;) (i32.const 1028) "*\00\00\00")
  (data $AnotherGreatSection (;2;) (i32.const 1032) "\07\00\00\00")
  (@custom "target_features" (after data) "\06+\0bbulk-memory+\16call-indirect-overlong+\0amultivalue+\0fmutable-globals+\0freference-types+\08sign-ext")
)
```
Without bulk-memory we get a .bss section:
```
(module $custom-section-name.bss.wasm
  (type (;0;) (func))
  (import "env" "memory" (memory (;0;) 2))
  (func $__wasm_call_ctors (;0;) (type 0))
  (global $__stack_pointer (;0;) (mut i32) i32.const 66576)
  (data $WowZero! (;0;) (i32.const 1024) "\00\00\00\00")
  (data $MyAwesomeSection (;1;) (i32.const 1028) "*\00\00\00")
  (data $AnotherGreatSection (;2;) (i32.const 1032) "\07\00\00\00")
  (data $.bss (;3;) (i32.const 1036) "\00\00\00\00")
  (@custom "target_features" (after data) "\05+\16call-indirect-overlong+\0amultivalue+\0fmutable-globals+\0freference-types+\08sign-ext")
)
```

The test is intended to test the use of the .bss section, and we only get a .bss section in this test if we disable bulk-memory.

Reply:
Hm, interesting, so we are using passive segments anytime bulk memory is enabled. That makes sense since this was all meant for threads, but is this actually an improvement in the non-threads case? i.e. maybe we don't want the extra explicit initialization code if we can help it?

edit: i just checked this out but I'll leave it here FTR: it looks like after linking we actually don't use passive segments unless threads is also enabled. so this seems like not a problem.